mnelab/mainwindow.py

```diff
@@ -291,10 +297,44 @@ def __init__(self, model):
         self.setAcceptDrops(True)
         self.data_changed()
 
+    def _sidebar_edit_event(self, edit):
+        """
+        Triggered when a data set in the sidebar is renamed.
+
+        Parameters
+        ----------
+        edit : PySide6.QtWidgets.QLineEdit
+            The text editor.
+        """
+        self.model.current["name"] = edit.text()
+
+    def _sidebar_move_event(self, parent, start, end, destination, row):
+        """
+        Triggered when an item in the sidebar is moved.
+
+        Parameters
+        ----------
+        parent : PySide6.QtCore.QModelIndex
+            ???
+        start : int
+            The source index of the item.
+        end : int
+            Same as start because the sidebar only allows single selection.
+        destination : PySide6.QtCore.QModelIndex
+            ???
+        row : int
+            The target index.
+        """
+        self.model.move_data(start, row)
+
     def data_changed(self):
         # update sidebar
-        self.names.setStringList(self.model.names)
-        self.sidebar.setCurrentIndex(self.names.index(self.model.index))
+        self.sidebar.clear()
+        self.sidebar.insertItems(0, self.model.names)
+        self.sidebar.setCurrentRow(self.model.index)
+        for it in range(self.sidebar.count()):
+            item = self.sidebar.item(it)
+            item.setFlags(item.flags() | Qt.ItemIsEditable)
```

Comment:
Why is this required on every call of this function? Can we make the items editable when they are first inserted?

Reply:
The sidebar is cleared and newly filled on every `data_changed`, because it doesn't "know" what happened to the `model`. I found a way to avoid this, but it requires a new connection from the `model` to its `view`, as this doesn't work via the `model.data_changed` decorator.

Reply:
What about `self.sidebar.model`?

Reply:
> So it's about syncing the entries in the model and the sidebar, right? Are you saying that the previous implementation also cleared everything and then added the new items with `self.names` (a `QStringList`)? Yes, it seems like it.

Yes & yes

> My original comment referred to the repeated `setFlags` call in the loop - is this not possible without a loop?

Apparently not :/

[That's](https://github.com/hofaflo/mnelab/commit/095548afe8fcd3cf18e851aab37ab8be16e16665) what my solution would look like... while it would be nice to not repopulate the sidebar all the time, I'm not sure if it's worth to give up the single connection from `Model` to `MainWindow` (in this direction) for that 🤔

Reply:
Right now, the methods in `Model` that need to trigger some change in the GUI are decorated with `data_changed`, which calls `Model.view.data_changed()`. That's the only place where a `Model` accesses its `view`. The proposed solution would introduce [two](https://github.com/hofaflo/mnelab/commit/095548afe8fcd3cf18e851aab37ab8be16e16665#diff-ea92b193a8d64b1c59aeebcb9d0f88a982171e429db17ad790d347941c2e3055R57) [new](https://github.com/hofaflo/mnelab/commit/095548afe8fcd3cf18e851aab37ab8be16e16665#diff-ea92b193a8d64b1c59aeebcb9d0f88a982171e429db17ad790d347941c2e3055R68) spots where `view` is accessed. While that isn't per se bad, I was thinking you kept the one-manages-all approach for a reason :D

Reply:
Ah, now I get it! Yes, I like the separation between the model and the view, even though it is not perfect. So yes, let's not introduce more imperfections for now...